src/videoCache.js

```diff
@@ -0,0 +1,213 @@
+/**
+ * This module interacts with the server used to cache video ad content to be restored later.
+ * At a high level, the expected workflow goes like this:
+ *
+ *   - Request video ads from Bidders
+ *   - Generate IDs for each valid bid, and cache the key/value pair on the server.
+ *   - Return these IDs so that publishers can use them to fetch the bids later.
+ *
+ * This trickery helps integrate with ad servers, which set character limits on request params.
+ */
+
+import { ajax } from './ajax';
+import { tryAppendQueryString } from './utils';
+
+const PUT_URL = 'https://prebid.adnxs.com/pbc/v1/put'
+const GET_URL = 'https://prebid.adnxs.com/pbc/v1/get?';
+const EMPTY_VAST_RESPONSE = '<VAST version="3.0"></VAST>'
+
+/**
+ * These are the properties required on a Bid in order to cache and retrieve it.
+ *
+ * @typedef {object} CacheableBid
+ * @property {string} vastUrl A URL which loads some valid VAST XML.
+ */
+
+/**
+ * Function which wraps a URI that serves VAST XML, so that it can be loaded.
+ *
+ * @param {string} uri The URI where the VAST content can be found.
+ * @return A VAST URL which loads XML from the given URI.
+ */
+function wrapURI(uri) {
+  // Technically, this is vulnerable to cross-script injection by sketchy vastUrl bids.
+  // We could make sure it's a valid URI... but since we're loading VAST XML from the
+  // URL they provide anyway, that's probably not a big deal.
+  return '<VAST version="2.0"><Ad id=""><Wrapper><AdSystem>prebid.org wrapper</AdSystem><VASTAdTagURI><![CDATA[' +
+    uri +
+    ']]></VASTAdTagURI><Impression></Impression><Creatives></Creatives></Wrapper></Ad></VAST>'
+}
+
+/**
+ * Wraps a bid in the format expected by the prebid-server endpoints, or returns null if
+ * the bid can't be converted cleanly.
+ *
+ * @param {CacheableBid} bid
+ */
+function toStorageRequest(bid) {
+  return {
+    'value': wrapURI(bid.vastUrl)
+  };
+}
+
+/**
+ * If the server responded successfully, then it gave us a bunch of objects with uuid properties.
+ * Abstract our background dependency away by renaming them to cacheId.
+ */
+function fromStorageResponse(response) {
+  return {
+    cacheId: response.uuid
+  };
+}
+
+/**
+ * A function which should be called with the results of the storage operation.
+ *
+ * @callback videoCacheStoreCallback
+ *
+ * @param {Error} [error] The error, if one occurred.
+ * @param {?string[]} uuids An array of unique IDs. The array will have one element for each bid we were asked
+ *   to store. It may include null elements if some of the bids were malformed, or an error occurred.
+ *   Each non-null element in this array is a valid input into the retrieve function, which will fetch
+ *   some VAST XML which can be used to render this bid's ad.
+ */
+
+/**
+ * A function which bridges the APIs between the videoCacheStoreCallback and our ajax function's API.
```

Comment:
A fair question... my belief is that each function, file, and API should make as much sense as possible on its own.

If you're looking at the `ajax` file, odds are that you want to make an HTTP request. Its API lets you set HTTP headers, considers any 200 to be a success, gives your callback function gets an HTTP response, etc.

If you're looking at the `videoCache` file, odds are that you want to cache a video somewhere. Its API lets you give it data to store, considers _any_ form of "cache miss" to be a failure, and gives your callback function the data it was seeking.

In between, there's code. "Data to store" needs to become a POST body. 200's need to be interpreted as JSON. Request details (URLs, headers, etc) can be chosen without bothering the caller who just wants to store their video somewhere.

Some of that code is responsible for transforming the callbacks. That's what I'm calling a "shim".